AdyenDropIn/Modules/Component/ComponentContainerViewModel.swift

```diff
@@ -45,14 +66,26 @@ internal class ComponentContainerViewModel: ComponentContainerViewModelProtocol
         component.viewController
     }
 
-    internal func didCancel() {
-        cancelHandler?(isRoot)
+    internal func cancel() {
+        handleCancellation()
     }
 
     // MARK: - Private
 
     private func setupComponent() {
         (component as? PaymentComponent)?.delegate = self
+        (component as? CardComponent)?.cardComponentDelegate = cardComponentDelegate
+        (component as? PartialPaymentComponent)?.partialPaymentDelegate = partialPaymentDelegate
+        (component as? PartialPaymentComponent)?.readyToSubmitComponentDelegate = self
+    }
+
+    private func setupActionComponent() {
+        actionComponent.delegate = self
```

Comment:
Why is component creation and setup are split? I'd make it either lazy var and create and setup on demand or call `setupActionComponent` right away after its creation. Otherwise current approach could open up a possibility to have at some point component created but not configured.

Reply:
Good point about creating it on demand, I haven't thought of it. I moved it to a computed property.

```swift
    private var actionComponent: AdyenActionComponent {
        let actionComponent = AdyenActionComponent(context: context)
        actionComponent.delegate = self
        actionComponent.presentationDelegate = self
        actionComponent.configuration.style = configuration.style.actionComponent
        actionComponent.configuration.localizationParameters = configuration.localizationParameters
        actionComponent.configuration.threeDS = configuration.actionComponent.threeDS
        actionComponent.configuration.twint = configuration.actionComponent.twint
        return actionComponent
    }
```